File: frontend/src/api/sqs.ts
import { SQSService } from "@bindings/bridge";
import type {
  SQSPublishInput,
  SQSPublishResult,
  SQSQueueInput,
} from "@bindings/bridge/models";
import { required } from "./client";

export type { SQSPublishInput, SQSPublishResult, SQSQueueInput };

/**
 * The SQS-only half of the surface.
 *
 * Reading queues is not here: a queue is a destination, and api/topic.ts
 * already answers the whole read side. What this carries is what the canonical
 * services cannot express - a create whose fields are durations and a redrive
 * policy, where TopicService.Create collects a broker address, two queue
 * counts and a permission string.
 */

/**
 * Declare a queue in the connection's region.
 *
 * Whether it is FIFO is decided by the name: SQS requires the .fifo suffix on
 * a FIFO queue and refuses it on a standard one, so the flag and the name have
 * to agree and a mismatch is refused before the request is sent.
 */
export const createQueue = (connID: number, input: SQSQueueInput): Promise<void> =>
  SQSService.CreateQueue(connID, input);

/**
 * Change an existing queue's settings.
 *
 * Only what the form sends is written. SQS replaces exactly the attributes it
 * is given, so a field left at zero keeps whatever is stored rather than
 * resetting it - which is what lets an edit change one setting alone.
 */
export const updateQueue = (connID: number, input: SQSQueueInput): Promise<void> =>
  SQSService.UpdateQueue(connID, input);

/**
 * Delete a queue and everything in it.
 *
 * There is no undo, and the name is refused for 60 seconds afterwards: a queue
 * recreated straight away fails, which is the service rather than the app.
 */
export const removeQueue = (connID: number, name: string): Promise<void> =>
  SQSService.RemoveQueue(connID, name);

/**
 * Discard everything the queue holds.
 *
 * The call returning is not the queue being empty. SQS purges asynchronously,
 * may also delete anything sent in the following minute, and may go on
 * delivering what was sent just before it for about as long. One purge per
 * queue per minute is allowed.
 */
export const purgeQueue = (connID: number, name: string): Promise<void> =>
  SQSService.PurgeQueue(connID, name);

/**
 * Send one body, or the same body several times, to one queue.
 *
 * Batched in tens, which is the service's own maximum, and a batch's entries
 * succeed and fail individually - so a send that stopped partway reports how
 * many the service took before it did.
 */
export const publish = (connID: number, input: SQSPublishInput): Promise<SQSPublishResult> =>
  SQSService.Publish(connID, input).then(required);

File: internal/bridge/sqs.go
package bridge

import (
	"context"

	sqsdriver "github.com/amigoer/mq-studio/internal/driver/sqs"
	sqsservice "github.com/amigoer/mq-studio/internal/service/sqs"
)

// SQSService is the renderer's entry point for the operations only Amazon SQS
// has. Listing and describing queues go through the canonical services; what
// is here is the rest.
type SQSService struct {
	service *sqsservice.Service
}

// NewSQSService wires the bridge to the service.
func NewSQSService(service *sqsservice.Service) *SQSService {
	return &SQSService{service: service}
}

// SQSQueueInput is a queue as the queue form collects it.
//
// Deliberately not TopicService.Create's shape. That one takes a broker
// address, two queue counts and a permission string - RocketMQ's vocabulary,
// of which an SQS queue has none. What a queue has instead is a set of
// durations the service enforces, and a redrive policy naming another queue.
//
// Every duration is zero when the form left it alone, which on an edit means
// "keep what is stored" rather than "set it to zero": SQS replaces exactly the
// attributes it is given, so an omitted one survives.
type SQSQueueInput struct {
	Name string `json:"name"`
	// FIFO is fixed at creation and has to match the .fifo suffix. It is sent
	// so a mismatch is refused with a message naming the field the user
	// actually filled in, rather than the attribute the service names.
	FIFO bool `json:"fifo"`

	VisibilityTimeoutSec int `json:"visibilityTimeoutSec"`
	DelaySec             int `json:"delaySec"`
	RetentionSec         int `json:"retentionSec"`
	MaxMessageBytes      int `json:"maxMessageBytes"`
	ReceiveWaitSec       int `json:"receiveWaitSec"`

	// DeadLetterQueue is another queue's name; the driver resolves its ARN.
	// Empty leaves the queue without a redrive policy.
	DeadLetterQueue string `json:"deadLetterQueue"`
	MaxReceiveCount int    `json:"maxReceiveCount"`

	// FIFO only. Ignored on a standard queue, where SQS refuses each of them
	// by an attribute name the form never drew.
	ContentBasedDeduplication bool   `json:"contentBasedDeduplication"`
	DeduplicationScope        string `json:"deduplicationScope"`
	FifoThroughputLimit       string `json:"fifoThroughputLimit"`
}

func (input SQSQueueInput) spec() sqsdriver.QueueSpec {
	return sqsdriver.QueueSpec{
		Name:                      input.Name,
		FIFO:                      input.FIFO,
		VisibilityTimeoutSec:      input.VisibilityTimeoutSec,
		DelaySec:                  input.DelaySec,
		RetentionSec:              input.RetentionSec,
		MaxMessageBytes:           input.MaxMessageBytes,
		ReceiveWaitSec:            input.ReceiveWaitSec,
		DeadLetterQueue:           input.DeadLetterQueue,
		MaxReceiveCount:           input.MaxReceiveCount,
		ContentBasedDeduplication: input.ContentBasedDeduplication,
		DeduplicationScope:        input.DeduplicationScope,
		FifoThroughputLimit:       input.FifoThroughputLimit,
	}
}

// CreateQueue declares a queue in the connection's region.
func (s *SQSService) CreateQueue(connID int, input SQSQueueInput) error {
	return s.service.CreateQueue(context.Background(), connID, input.spec())
}

// UpdateQueue changes an existing queue's settings. Whether a queue is FIFO is
// not among them: that is fixed at creation and spelled in its name.
func (s *SQSService) UpdateQueue(connID int, input SQSQueueInput) error {
	return s.service.UpdateQueue(context.Background(), connID, input.spec())
}

// RemoveQueue deletes a queue and everything in it. There is no undo, and SQS
// refuses the name for 60 seconds afterwards.
func (s *SQSService) RemoveQueue(connID int, name string) error {
	return s.service.RemoveQueue(context.Background(), connID, name)
}

// PurgeQueue discards everything the queue holds.
//
// The call returning is not the queue being empty: SQS purges asynchronously,
// may also delete anything sent in the following minute, and may keep
// delivering what was sent before it for about as long.
func (s *SQSService) PurgeQueue(connID int, name string) error {
	return s.service.PurgeQueue(context.Background(), connID, name)
}
